### ISSUE TITLE
Create easy way to schedule materiazed value side effects

### ISSUE THREAD

**Issue by acjay** [akka/akka#20184]
I’m attempting to subscribe an actor to `eventStream` within the `Graph` I’m creating for a `Flow` as the handler for `handleWebSocketMessages`, but for the life of me, I can’t figure out how to get the `actorRef` from `Source.actorRef`. How can I operate on its materialized value within the graph?

This `Source.actorRef` is feeding into a `Merge`, and it’s not clear to me what the `b.materializedValue` of my `Graph`’s `builder` will be, and the compiler doesn’t seem to know either.

But I feel like, even if I knew the answer to this question, it's a super awkward interface anyway. It seems like it would be better to be able to attach some side effect to the materialization of the `Source` directly, or perhaps any partial `Graph`. that way i could do something like:

```
val actorInputSource = Source.actorRef[_](???, ???)
  .whenMaterialized(actorRef => system.eventStream.subscribe(actorRef, classOf[EventType]))
```

This pattern feels like it could potentially be useful anywhere there's a desire to shunt a materialized value out of a `Graph`.

**Comment by acjay** [comment]
Alternatively, https://github.com/akka/akka/blob/0d8dd3c7114ed5258208aaed3575460a056876bb/akka-stream/src/main/scala/akka/stream/impl/ActorRefSourceActor.scala#L27-L109 basically does exactly what I'm looking for, and potentially would serve the same purpose. Would it potentially make sense to make this public? I guess the only practical difference would be the order of instantiation between the `Flow` and the `Actor`.

**Comment by drewhk** [comment]
> How can I operate on its materialized value within the graph?

In a graph (if you mean `GraphDSL`) you can access the materialized value inside the graph as a source: http://doc.akka.io/docs/akka/2.4.2/scala/stream/stream-graphs.html#Accessing_the_materialized_value_inside_the_Graph

> This Source.actorRef is feeding into a Merge, and it’s not clear to me what the b.materializedValue of my Graph’s builder will be, and the compiler doesn’t seem to know either.

Please give a reproducer. From that sentence I have no clue what you have attempted, what you have expected to happen, and what happened instead. If you give a code snippet then we will likely be able to help here.

> like it would be better to be able to attach some side effect to the materialization of the Source directly, 

You can use `mapMaterializedValue` for that, it already exists.

**Comment by acjay** [comment]
> In a graph (if you mean GraphDSL) you can access the materialized value inside the graph as a source

I pored over those docs, but still wasn't able to make sense of how that works. Does this only 

> You can use mapMaterializedValue for that, it already exists.

Ah, excellent. I can adapt this to what I'm trying to do. It seems like it would be idiomatic to have a side-effect-centric alias still.

> Please give a reproducer.

Here's a simplified version of what I have today:

```
// Defining a flow for a websocket
Flow.fromGraph(GraphDSL.create() { implicit b =>
    import GraphDSL.Implicits._

    // One flow of outgoing messages is the flow from the incoming message,
    // through a request procesor.
    val requestResponseFlow = Flow.fromFunction(requestResponseProcessor).mapAsync(1)(identity)

    // The other flow of outgoing messages is the flow from the actor that
    // listens for applicable lot updates.
    val updateBroadcastActorProps = UserUpdateListener.props
    val updateBroadcastEventSource = Source.actorPublisher[OutgoingWebsocketEvent](updateBroadcastActorProps)

    // ^ This is an `ActorPublisher` that does nothing but buffer messages
    // received from subscribing to the eventStream. What I really want is:
    val updateProcessorFlow = Flow.fromFunction(updateProcessor)
    val updateBroadcastEventSource = Source.actorRef[UpdateMesage](10, OverflowStrategy.dropHead).via(updateProcessorFlow)
    val updateActorRef = ???
    system.eventStream.subscribe(updateActorRef, classOf[UpdateMessage])

    // Merge the two flows of outgoing messages.
    val processedInflow = b.add(decodeMessage.via(requestResponseFlow))

    val merge = b.add(Merge[OutgoingWebsocketEvent](2))
    processedInflow ~> merge.in(0)
    updateBroadcastEventSource ~> merge.in(1)

    // Encode the flows for transmission back to the client.
    val output = b.add(Flow[OutgoingWebsocketEvent].via(encodeMessage))
    merge ~> output

    FlowShape(requestResponse.in, output.out)
  })
```

**Comment by acjay** [comment]
`mapMaterializedValue` worked. I'll go ahead and close, but I do think that a much more obvious way of getting a handle on the `ActorRef` for `Source.actorRef` would be desirable. Perhaps an optional 3rd parameter `withActorRefDo: ActorRef => Unit = _ => ()`. If this would be amenable, I'd be glad to make a PR, because I think it would be a simple change.

**Comment by drewhk** [comment]
> I pored over those docs, but still wasn't able to make sense of how that works. Does this only

My guess is that the graph you created had no materialized value. The modifications should roughly look like:

```
// Must pass the Graph you want to import *and* use its materialized value as an argument to `create`. This is necessary because otherwise the type system cannot infer any materialized value type since a `b.add()` call cannot introduce type information on the overall graph
Flow.fromGraph(GraphDSL.create(sinkToBeImported) { implicit b => 
    // You will get then the imported *instance* of that graph as an argument to your builder block
    importedSink =>
    import GraphDSL.Implicits._
    b.materializedValue // Now it will has the proper type which is the materialized value type of sinkToBeImported
```

The above pattern is what we usually refer to as *static* import as it is done via arguments to the builder, in comparison to the `builder.add()` method which is dynamic (can be done in a for-loop) but cannot change the materialized value type of the resulting graph.

**Comment by drewhk** [comment]
> but I do think that a much more obvious way of getting a handle on the ActorRef for Source.actorRef would be desirable

We are rather strict with adding sugar to the library as things tend to end up as a kitchen sink of stuff. There are opposing forces here:
 - too little sugar means frustrated users, not finding the "simple thing" they want because it exists only as a *combination* of existing things
 - too much sugar means frustrated users, not finding the "simple thing" they want because it is hidden somewhere in a huge pile of existing, specialized things (where names are never "right" given any two nonequal programmer)

We are on the side of keeping only small amount of sugar around, as once you get the feel for it you will be able to build stuff easily as your knowledge *generalizes* compared to knowledge of an ad hoc collection of highly specific solutions.

`mapMaterializedValue` is such a building block, and while it is possible to add a sugar on top of it, the duplication will cause confusion for a different set of programmers (we considered adding a side-effecting alias in the past, but rejected the idea in the end).

tl;dr; We only add sugar if enough users cry :D

**Comment by drewhk** [comment]
OTOH you can add a cookbook recipe describing this pattern, we didn't grow it since a while so new additions are welcome.

**Comment by acjay** [comment]
> My guess is that the graph you created had no materialized value. 

That's hella confusing :)

Yeah, so I think in my case, I'd want to add the `Source.actorRef[UpdateMesage](10, OverflowStrategy.dropHead)` as a parameter to `create`.

> tl;dr; We only add sugar if enough users cry :D

I hear you on the balance, but don't think materialized value feels like a natural way to access the `ActorRef`. `Sink.actorRef`, by contrast, doesn't require it. Maybe it feels weird to me because I'm not knee deep in Akka Streams and using materialized values anywhere else. I just wanted to make a simple websocket server with async push, and that's the only place I'm using Akka Streams in this entire project. Now, not only do I have to know this, but it has to be maintainable for less experienced Scala programmers -- e.g. the entire rest of my organization.

As a piece of feedback on the API, it took me *hours* of docs reading and several Gitter questions just to figure this out. I had to read the entirety of the first 7 docs -- with a digression to the Integration doc, before I was able to retire my hand-rolled `ActorSubscriber` and `ActorPublisher`.

I can and will document my approach, but I stand by my assessment that even with a concise solution, the current situation is problematic.
